pkg/k8s/portforward.go

```diff
@@ -162,6 +164,37 @@ func (pf *PortForward) Run() error {
 	return fw.ForwardPorts()
 }
 
+// Init creates and runs a port-forward connection.
+// This function blocks until the connection is established, in which case it returns nil.
+func (pf *PortForward) Init() error {
+	log.Debugf("Starting port forward to %s %d:%d", pf.url, pf.localPort, pf.remotePort)
+
+	failure := make(chan error, 1)
+
+	go func() {
+		if err := pf.Run(); err != nil {
+			failure <- err
+		}
+
+		if _, ok := <-pf.stopCh; ok {
```

Comment:
I'm having trouble figuring out the semantics of stopCh.  Is it the case that this channel is written to when the port-forward fails and the channel is closed when the port-forward is explicitly stopped?  

This line of code seems to wait until the channel is closed or written to, and then closes the channel if it hasn't been.  What is the purpose of this?

What happens if the channel fails after being stopped?  Could the portforward attempt to write a message to the closed channel and panic?

Reply:
I wonder if this will end up competing with the `select` statement in `portforward.ForwardPorts()`, as seen [here](https://github.com/kubernetes/client-go/blob/2c6e35a5b9cf292580c2d8ecb7d141572997d3cd/tools/portforward/portforward.go#L230-L234).

Reply:
Right, `stopCh` is only used as a broadcast mechanism to signal we want the connection to end (by closing the channel). You can see in client-go's [portforward.go](https://github.com/kubernetes/client-go/blob/2c6e35a5b9cf292580c2d8ecb7d141572997d3cd/tools/portforward/portforward.go) its type is always `<-chan` so it doesn't get written to. For the same reason, there is no competition when reading from it; once it's closed all its readers will unblock.

@siggy That's a nice idiom. I had the same intention but I got the meaning of `_, ok := <- ch` wrong.